### ISSUE TITLE
Connection to test server cannot be established

### ISSUE THREAD

**Issue by Oremuss** [rtr-nettest/open-rmbt#8]
Hello guys, me again :)

Now when my servers are working I am trying to make them work with Android app. 

There is no specific manual so I have looked into code and found some possible places where should I put my IP address of Control Server. Possibly **defaults.xml** in `res/values/defaults/` and **`res/values/xml/preferences_dev.xml`**. I have changed in defaults value of mapserver to my server IP address and also controlServer to the same server IP (they are all on 1 notebook as solved in my previous issue). 
Also I have left empty strings **`default_control_host_ipv4_only` , `default_control_host_ipv6_only ` , `default_control_check_ipv4_url` and `default_control_check_ipv6_url`.** Is it necessary to fill that strings?

 After building app and running it, when I press start button,  app  showes me error:**` Connection to test server cannot be established`**. 

**Is there any more settings which needs to be set?** 
What about key needed by database? In databse table test_server I have put to the column "key" some number, but nothing else. Now there is some uuid and my key, should I put there something esle or is it handled by the app?

**Comment by cproof** [comment]
Hi,

can you please paste the relevant log output (`/var/log/tomcat8/catalina.out` and the output of `logcat`/Android Studio) in the ticket?

You're usually all set, once you entered this values and entered the test server address in the DB.

**Comment by Oremuss** [comment]
Here is my catalina.out. I hope it will help to identify error.
[catalina.out.txt](https://github.com/rtr-nettest/open-rmbt/files/3906335/catalina.out.txt)
I dont have logcat now, because it is on my desktop and I only have my ntb with me and I am testing it with direct app to phone, not emulating now :)

**Comment by cproof** [comment]
That's really strange. It points to a problem where the client identifier that the app registers with is not contained in the ControlServer config (https://github.com/rtr-nettest/open-rmbt/blob/master/RMBTControlServer/src/at/rtr/rmbt/res/Configuration.properties#L17). I'll look into that.

**Comment by Oremuss** [comment]
Hmm like I said I am just begginer with this andstill learning, but maybe I did something wrong on the way. RIgh now if you try to connecto to `c1ca47d2.ngrok.io/RMBTControlServer/version` you can see my server. At least for next 2 hours will be available if you want try something, maybe some hidden feature I dont know about :) maybe it will help, anyway I am open to suggestion what to do :).

**Comment by cproof** [comment]
Hm...
You can try adding some sort of logging output in https://github.com/rtr-nettest/open-rmbt/blob/master/RMBTControlServer/src/at/rtr/rmbt/controlServer/SettingsResource.java#L97 (print "typeId", as this seems to not be contained in https://github.com/rtr-nettest/open-rmbt/blob/master/RMBTControlServer/src/at/rtr/rmbt/res/Configuration.properties#L17)

**Comment by Oremuss** [comment]
Ill do it, just to be clear, I should add to next line 98 (or 99 to that if condition)  `System.out.println(typeId); ` build server again, put WAR file to tomcat and restart tomcat8 right?

**Comment by cproof** [comment]
Yes.
However, its probably even easier to add a `System.out.println(entity);` directly in :55. This would make for even easier debugging.

**Comment by Oremuss** [comment]
Roger roger :) I'll make it and result should be here in couple of minutes ( If I dont get any error on the way ).

**Comment by Oremuss** [comment]
[catalina_out1.txt](https://github.com/rtr-nettest/open-rmbt/files/3906386/catalina_out1.txt)
Okay, round two.

**Comment by cproof** [comment]
TY, that helped 👍 

it seems that for some reason the 
```sql
INSERT INTO public.client_type (uid, name) VALUES (2, 'MOBILE');
```
from `rmbt_init.sql` was not executed in your database. 
Did you execute both `rmbt.sql` and `rmbt_init.sql` as outlined in the `README`?

If so, could you try a  `SELECT * FROM client_type;`? During the registration process, the client identifies itself (in your case, by its name `Open-RMBT` and type `MOBILE`). These identifiers are then looked up from the `Configuration.properties` and the database. If not found, the error message you're seeing is returned.

**Comment by Oremuss** [comment]
I did execute both, rmbt.sql and laso rmbt_sql, but now I made `SELECT * FROM client_type;` and it is empty as you pointed out.

**Comment by cproof** [comment]
In that case (`client_type` exists, but is empty) maybe try to again run `cat rmbt_init.sql | psql rmbt -1`.
That's really strange. We tried the whole set-up process ourselfes yesterday (and actually fixed some bugs in the `rmbt.sql`), but it's strange that the `rmbt_init.sql` failed on your system. Did you get any error message when running it?

> Hmmm, should I delete database and remake it again?    

That's probably the cleanest solution.

**Comment by Oremuss** [comment]
Thats the weird part, I didnt get any error here. Ill trz first reinit this part `cat rmbt_init.sql | psql rmbt -1`.

**Comment by Oremuss** [comment]
OK so now I rerun it all again and everything seems to be in place as it suppose to be. But, and here we go again, I have tried test on phone, but still failed, now it looks like some problem with QoSTestRequest...., sigh, I feel like I am making up problems.

**Comment by cproof** [comment]
Can you again post the `catalina.out`, if there is any error log? I may then come back to you tomorrow / Monday.
After we get it running, I'm sure we can improve the documentation of this project in case we're missing something 👍

**Comment by Oremuss** [comment]
Sure thing, but I will for sure try again everything from scratch if I somehow miss something and in both cases( succes or fail ) I will report my progres, possibly(or hopefully) close this issue :). Thank you very much for your help, I appreciate it very much.

**Comment by Oremuss** [comment]
Hello, so here is  catalina.out : 
[catalina_out2.txt](https://github.com/rtr-nettest/open-rmbt/files/3906497/catalina_out2.txt)

**Comment by cproof** [comment]
Hi,
that log looks good!
Did you set up the [rmbt-server](https://github.com/rtr-nettest/rmbt-server/) so that also the speed test can be performed?
If so, also `"test_server_port" : 0` seems strange (which is a setting from table `test_server`)

**Comment by Oremuss** [comment]
Well I didnt set up that, somehow I tought if it doesnt belong to open-rmbt folder it deosnt have to be installed :). Ok so rmbt-server and also rmbtws client has to set up in my laptop? I will look into that. I definitly reccomend to mention that in README for "begginers" like me :)
 Like I said in beggining, I am no expert in this area, primary I am C embedded developer but trying to learn new things in my free time.

**Comment by cproof** [comment]
If you only want to use the Android app, you only need `rmbt-server`, as `rmbtws` is the speed test client used in websites. And `rmbt-server` is way easier to set up than the server/db backend in this repo. Log output from  `rmbt-server` is either directly in `stdout` or  `/var/log/syslog`.
Just be sure to supply the keys and values used for configuring your `rmbt-server` also in the `test_server` table so that the Android client can authenticate itself.

**Comment by Oremuss** [comment]
OK I'll do it now, I will set up rmbt-server and try to run app with it, but I have no `"test_server_port" : 0` in `test_server` table.

**Comment by Oremuss** [comment]
Hmm I am little bit confused with those keys, which key should be put where? The secret.key should be put to database `test_server` table to "key" ?? 
[https://ctrlv.sk/shots/2019/11/30/SZnj.png](url)


You can see in image, my `test_server` table and key has value same as value. Is it correct? Also the port is suppose to be 80 in table and 8081 if I run server? ( I used default from makefile).

**Comment by cproof** [comment]
Does not matter, one of them (every line in the `secret.key` is a key that is accepted by the server). E.g the key starting with `4pH8q4R...`
Also, add port `8082` to `port_ssl`, and port `8081` to `port`. Then, you're good to go.

**Comment by Oremuss** [comment]
One step closer but still not it wouldnt be me if it will be that easy :D Now succesfully connected, 12% test and suddenly error again.

**Comment by cproof** [comment]
This behaviour is typical when the test server cannot be reached.
When accessing the server adress directly, i.e. https://c1ca47d2.ngrok.io:8082 in a browser or via curl you should get  `RMBTv1.2.1`

**Comment by Oremuss** [comment]
Accesing directly from URL doesnt work, it retrun This site cant be reached. I will have to look into that.

**Comment by cproof** [comment]
> So when I run rmbtd -l 8081 -L 8082 -c server.crt -k server.key -D it should be visible right after start when I see url localhost:8082 ??

Basically, yes.

I just tried it myself, while running
```bash
./rmbtd -l 8081 -L 8082 -k server.key -c server.crt
```

I get output from a separate shell
```bash
> curl -k https://localhost:8082/
RMBTv1.2.1
ACCEPT TOKEN QUIT
```

**Comment by Oremuss** [comment]
Hmmm same thing here.

**Comment by cproof** [comment]
That's actually fine. `rmbt` without the `-w` flag does not expect HTTP input, but [expects the token](https://www.netztest.at/doc/#phase_2_example) Try through the app again, it should hopefully work now.

**Comment by Oremuss** [comment]
The log is created when I tried it with app ( and already working server returning
`RMBTv1.2.1 ACCEPT TOKEN QUIT` ). Shouldnt I somehow tell to the app about port 8081 or 8082 ?

**Comment by cproof** [comment]
The information about the port should automatically be communicated to the app via the ControlServer, i.e. you have to set field `port_ssl` to `8082` in the `test_server`-table.

In `catalina.out` from yesterday you hade the line
```
Test request from 127.0.0.1 completed (602 ms)
{"client_remote_ip":"127.0.0.1","test_uuid":"bf1f8e2d-5394-472b-a4e7-554efb87c998","result_url":"http://c1ca47d2.ngrok.io/RMBTControlServer/result","result_qos_url":"http://c1ca47d2.ngrok.io/RMBTControlServer/resultQoS","test_duration":"7","test_server_name":"OpenRMBT Server (Vienna)","test_wait":0,"test_server_address":"c1ca47d2.ngrok.io","test_numthreads":"3","error":[],"test_server_port":0,"error_flags":[],"open_test_uuid":"O4299449f-900d-4c63-bfe7-b3a8fc1231ed","test_server_type":"RMBT","test_server_encryption":true,"test_token":"4299449f-900d-4c63-bfe7-b3a8fc1231ed_1575067441_j5fDn8CLW3pKvDbA0wXxA/DF5pU=","test_numpings":"10","test_id":3} 
```
All configuration for the client is transmitted via the `test_server_port`, `test_server_address` and `test_server_encryption` JSON-Fields

**Comment by Oremuss** [comment]
What about that `"error":[] ` after  `"test_numthreads":"3"` , isnt it somehow wrong because of it?

**Comment by cproof** [comment]
>  error field reproting only errors and its emptz because of there is none error

Yes. No errors, so an empty array.

**Comment by Oremuss** [comment]
Ill try now reinstall app, restart tomcat service, restart server and report catalina.out. because I am running out of ideas at the moment.

**Comment by Oremuss** [comment]
Well according to report from catalina you are right, it looks like QoS looks good, but it stops in 12% and wont start Ping and just reports error as I posted screenshot already today, from curiosity I will try different phone with android 9, this one has android 6, but I think it deosnt have anything to do with this issue.

**Comment by cproof** [comment]
Can it be some problem or configuration issue with ngrok.io? Maybe you need to bind to some specific IP? Did you set in ngrok.io that this specific service is not HTTP but plain TCP?

At least if I try to connect to the server mentioned in your `catalina_out5`, I get a timeout: https://c1ca47d2.ngrok.io:8082

**Comment by Oremuss** [comment]
Hmm out of the otion actually it might be, when I set ngrok I use command `./ngrok http 8080`.

**Comment by cproof** [comment]
Could try something along the lines of `ngrok tcp 8082`? (I'm not familiar with ngrok, just copied from their documentation).
On a sidenode: If you're based in Europe, you ISP [is not allowed to block ports](https://ec.europa.eu/digital-single-market/en/policies/open-internet) without justification. If it is, your local [telecom regulator](https://en.wikipedia.org/wiki/List_of_telecommunications_regulatory_bodies) might be interested.

**Comment by Oremuss** [comment]
That would be problem as I can use ( as free user) only one port which is 8080, now if I turn that off and I will allow onlz 8082,  I will try to solve it, I gues it is the last thing that is blocking succesfull test. (I really hope so). I will try to bypass it.

**Comment by Oremuss** [comment]
Now I have made with ngrok, three connections to ports http 8080, tcp 8081 and 8082. Now curl works only with port 8081, port 8082 returns `Empty reply from server`. But it doesnt solve problem, becuae now I have three different IP adresses one is as ususall hhtp ngrok, and those two are totally different adresses I would say dead end. I need somehow to call my provider to enable ports for forwarding, it would be much easier :).

**Comment by Oremuss** [comment]
After couple of days beeing busy I have now some time for testing :) . Now I have managed to make VPN connection and now my VPN IP has open ports, I asked friend to curl my server `curl - k http://ip_addres:8081` succesfully and also port 8082, version, so with full hope I have  replaced my ngrok addres with just IP in database, changed in defaults in APP but again Failed to connect to server. 
Now beside this problem I have also long term poblem with category map which I noticed long time ago, but it was put aside. When I click on Map, the app always crash, it is crashing in emulator and also in my both phones with android 9 and Android 6. Is somehow my map server in wrong configuration?
